File: LowRes NX Coder/LowRes NX Coder/App/AppController.m
//
//  AppController.m
//  Pixels
//
//  Created by Timo Kloss on 3/3/15.
//  Copyright (c) 2015 Inutilis Software. All rights reserved.
//

#import "AppController.h"
#import "UIViewController+LowResCoder.h"
#import "HelpContent.h"
#import "TabBarController.h"

NSString *const ShowPostNotification = @"ShowPostNotification";
NSString *const UpgradeNotification = @"UpgradeNotification";
NSString *const ImportProjectNotification = @"ImportProjectNotification";


@implementation TempProject
@end


@implementation AppController

+ (AppController *)sharedController
{
    static AppController *sharedController = nil;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        sharedController = [[self alloc] init];
    });
    return sharedController;
}

- (instancetype)init
{
    if (self = [super init])
    {
        NSURL *url = [[NSBundle mainBundle] URLForResource:@"manual" withExtension:@"html" subdirectory:@"docs"];
        _helpContent = [[HelpContent alloc] initWithURL:url];
        
        _bootTime = CFAbsoluteTimeGetCurrent();
    }
    return self;
}

- (void)showAlertWithTitle:(NSString *)title message:(NSString *)message
{
    UIViewController *vc = self.tabBarController.selectedViewController;
    if (vc.presentedViewController)
    {
        vc = vc.presentedViewController;
    }
    
    UIAlertController *alert = [UIAlertController alertControllerWithTitle:title message:message preferredStyle:UIAlertControllerStyleAlert];
    [alert addAction:[UIAlertAction actionWithTitle:@"OK" style:UIAlertActionStyleDefault handler:nil]];
    
    [vc presentViewController:alert animated:YES completion:nil];
}

- (BOOL)handleOpenURL:(NSURL *)url
{
    /*
    if ([[url scheme] isEqualToString:@"lowrescoder"])
    {
        NSMutableDictionary *params = [NSMutableDictionary dictionaryWithParamsFromURL:url];
        NSString *postId = params[@"lccpost"];
        if (postId)
        {
            self.shouldShowPostId = postId;
            [[NSNotificationCenter defaultCenter] postNotificationName:ShowPostNotification object:self];
        }
        return YES;
    }
    else
    */
    if (url.isFileURL)
    {
        // load text file for new project
        NSError *error = nil;
        NSString *fileText = [NSString stringWithContentsOfURL:url encoding:NSUTF8StringEncoding error:&error];
        if (fileText)
        {
            NSString *name = url.lastPathComponent;
            if (url.pathExtension.length > 0)
            {
                name = [name substringToIndex:name.length - url.pathExtension.length - 1];
            }
            TempProject *tempProject = [[TempProject alloc] init];
            tempProject.name = name;
            tempProject.sourceCode = fileText;
            self.shouldImportProject = tempProject;
            [[NSNotificationCenter defaultCenter] postNotificationName:ImportProjectNotification object:self];
        }
        else
        {
            NSLog(@"error: %@", error.localizedDescription);
        }
    }
    return NO;
}

@end

File: LowRes NX Coder/LowRes NX Coder/App/AppController.swift
//
//  AppController.swift
//  LowRes NX Coder
//
//  Created by Timo Kloss on 22/02/2019.
//  Copyright © 2019 Inutilis Software. All rights reserved.
//

import UIKit
import StoreKit

@objc class AppController: NSObject {
    
    private static let hasDontatedKey = "hasDontated"
    private static let isSafeScaleEnabledKey = "isSafeScaleEnabled"
    private static let numRunProgramsThisVersionKey = "numRunProgramsThisVersion"
    private static let lastVersionKey = "lastVersion"
    private static let lastVersionPromptedForReviewKey = "lastVersionPromptedForReview"
    private static let userIdKey = "userIdKey"
    private static let usernameKey = "usernameKey"
    
    @objc static let shared = AppController()
    
    @objc weak var tabBarController: TabBarController!
    
    @objc let helpContent: HelpContent
    @objc let bootTime: CFAbsoluteTime
    
    private var webSource: WebSource?
    
    var hasDontated: Bool {
        get {
            return UserDefaults.standard.bool(forKey: AppController.hasDontatedKey)
        }
        set {
            UserDefaults.standard.set(newValue, forKey: AppController.hasDontatedKey)
        }
    }
    
    var isSafeScaleEnabled: Bool {
        get {
            return UserDefaults.standard.bool(forKey: AppController.isSafeScaleEnabledKey)
        }
        set {
            UserDefaults.standard.set(newValue, forKey: AppController.isSafeScaleEnabledKey)
        }
    }
    
    var numRunProgramsThisVersion: Int {
        get {
            return UserDefaults.standard.integer(forKey: AppController.numRunProgramsThisVersionKey)
        }
        set {
            UserDefaults.standard.set(newValue, forKey: AppController.numRunProgramsThisVersionKey)
        }
    }
    
    private(set) var userId: String? {
        get {
            return UserDefaults.standard.string(forKey: AppController.userIdKey)
        }
        set {
            UserDefaults.standard.set(newValue, forKey: AppController.userIdKey)
        }
    }
    
    private(set) var username: String? {
        get {
            return UserDefaults.standard.string(forKey: AppController.usernameKey)
        }
        set {
            UserDefaults.standard.set(newValue, forKey: AppController.usernameKey)
        }
    }
    
    var currentVersion: String {
        let version = Bundle.main.object(forInfoDictionaryKey: kCFBundleVersionKey as String) as! String
        return version
    }
    
    private override init() {
        let url = Bundle.main.url(forResource: "manual", withExtension: "html", subdirectory:"docs")!
        helpContent = HelpContent(url: url)
        
        bootTime = CFAbsoluteTimeGetCurrent()
        
        super.init()
        
        let lastVersion = UserDefaults.standard.string(forKey: AppController.lastVersionKey)
        if currentVersion != lastVersion {
            numRunProgramsThisVersion = 0
            UserDefaults.standard.set(currentVersion, forKey: AppController.lastVersionKey)
        }
    }
    
    func requestAppStoreReview() {
        let lastVersionPromptedForReview = UserDefaults.standard.string(forKey: AppController.lastVersionPromptedForReviewKey)
        if currentVersion != lastVersionPromptedForReview {
            if #available(iOS 10.3, *) {
                SKStoreReviewController.requestReview()
            }
            UserDefaults.standard.set(currentVersion, forKey: AppController.lastVersionPromptedForReviewKey)
        }
    }
    
    func runProgram(_ webSource: WebSource) {
        if tabBarController != nil {
            showProgram(webSource)
        } else {
            self.webSource = webSource
        }
    }
    
    @objc func checkShowProgram() {
        if let webSource = webSource {
            showProgram(webSource)
            self.webSource = nil
        }
    }
    
    private func showProgram(_ webSource: WebSource) {
        let storyboard = UIStoryboard(name: "LowResNX", bundle: nil)
        let vc = storyboard.instantiateInitialViewController() as! LowResNXViewController
        vc.webSource = webSource
        
        if tabBarController.presentedViewController != nil {
            tabBarController.dismiss(animated: false, completion: nil)
        }
        tabBarController.present(vc, animated: true, completion: nil)
    }
    
    func didLogIn(userId: String, username: String) {
        self.userId = userId
        self.username = username
    }
    
    func didLogOut() {
        self.userId = nil
        self.username = nil
    }
    
}
